_posts/2019-02-20-refactoring-state-pattern.markdown

```diff
@@ -0,0 +1,114 @@
+---
+layout: post
+title:  "Refactoring with Design Patterns - The State Pattern"
+date: 2019-02-20 10:00:00
+categories: ["code-refactor", "design-patterns"]
+author: "cleiviane"
+---
+
+In this series of [code refactoring](https://www.ombulabs.com/blog/tags/code-refactor) posts we are discussing how design patterns can be used to make our [Ruby](https://www.ruby-lang.org/en/) code beautiful, maintainable and clean.
+
+Today I want to talk about a pattern that can be very useful when we need to control the flow of a set of events of our objects: *The State Pattern* a.k.a *Finite State Machine*.
+
+As a developer it is common to see objects changing of states. At the beginning managing the state of an object can be as simple as having some boolean attributes where you can check if the object is in state A or B. But when the complexity increases you can end up with a number of states that are difficult to manage without breaking the [SOLID principles](http://butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod). That is where we can benefit from the elegant solution the State Pattern gives to us.
+
+<!--more-->
+
+## The State Pattern
+
+According to the [Refactoring Guru](https://refactoring.guru/design-patterns/state): <strong>_"The main idea of the State pattern is that, at any given moment, there’s a finite number of states which an object can be in. Within any unique state, the program behaves differently, and the program can be switched from one state to another instantaneously."_</strong>
+
+Think of the solution as a [finite sate machine](https://brilliant.org/wiki/finite-state-machines/) where you can control your objects making sure that they can be in only one state at a time, so the objects will transition from one state to another to perform a set of different actions.
+
+Here at Ombulabs a good example of when we need to use state pattern is in one of our products, [Ombushop](https://secure.ombushop.com). It's an e-commerce platform where users can create their customized store and sell products online.
```

Comment:
@cleicar I think the link should be https://www.ombushop.com

https://secure.ombushop.com is for the admin panel.